# Pandas CheatSheet for Data Exploration

This cheat sheet is perfect for anyone transitioning from Excel to Python and Pandas, especially if you’re used to Excel's intuitive UI for data handling. Before my switch to Python, one of my main challenges was adapting to data manipulation in a more code-driven environment. If you’re familiar with concepts like indexing, grouping, aggregating, and cleaning data in Excel, this guide will help you translate those skills into Python using Pandas, focusing on the core DataFrame and Series objects.

Most of the examples are concise, often requiring just one line of code, which optimizes your workflow.

For practice, I’m using a fun dataset from Kaggle: Superhero Dataset. https://www.kaggle.com/datasets/claudiodavi/superhero-set

I also give all credit to 100 Pandas Puzzles for inspiring many of the solutions and examples included. 
https://github.com/ajcr/100-pandas-puzzles/blob/master/100-pandas-puzzles-with-solutions.ipynb


## Importing pandas and Data 


**1.** Import pandas and numpy under the alias `pd`.

In [265]:
import pandas as pd
import numpy as np

 Load the superhero dataset (assuming it's stored locally after being downloaded from Google Drive) in this case please just replace the id into 'https://drive.google.com/uc?id=*ID*' but If you want to upload the file locally use the option 2:
 
 *df = pd.read_csv("../input/telecom_churn.csv")*
 
 *df.head()*

We define our Dataset here as *"superherodata*"* and Read, and Display the first few rows

In [266]:
#Option 1 in googledrive
superherodata = 'https://drive.google.com/uc?id=1MR_wKUf50_stiiXI747l99C3tRMgIp4y'
df = pd.read_csv(superherodata)
df.head()

Unnamed: 0         name Gender Eye color               Race Hair color  \
0           0       A-Bomb   Male    yellow              Human    No Hair   
1           1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   
2           2     Abin Sur   Male      blue            Ungaran    No Hair   
3           3  Abomination   Male     green  Human / Radiation    No Hair   
4           4      Abraxas   Male      blue      Cosmic Entity      Black   

   Height          Publisher Skin color Alignment  Weight  
0   203.0      Marvel Comics          -      good   441.0  
1   191.0  Dark Horse Comics       blue      good    65.0  
2   185.0          DC Comics        red      good    90.0  
3   203.0      Marvel Comics          -       bad   441.0  
4   -99.0      Marvel Comics          -       bad   -99.0

## DataFrame basics

### Let’s have a look at data dimensionality, feature names, and feature types.


Display a summary of the basic information about this DataFrame and its data (*hint: there is a single method that can be called on the DataFrame*). We use this to understand out dataset, the shape its just the number of columns and rows, 


In [267]:
print(df.shape)

(734, 11)


From the output, we can see that the table contains 734 rows and 11 columns.

Now let's try printing out column names using columns:

In [268]:
print(df.columns)

Index(['Unnamed: 0', 'name', 'Gender', 'Eye color', 'Race', 'Hair color',
       'Height', 'Publisher', 'Skin color', 'Alignment', 'Weight'],
      dtype='object')


Let's use the  info() method to output some information about the dataframe: 

In [269]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 734 entries, 0 to 733
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  734 non-null    int64  
 1   name        734 non-null    object 
 2   Gender      734 non-null    object 
 3   Eye color   734 non-null    object 
 4   Race        734 non-null    object 
 5   Hair color  734 non-null    object 
 6   Height      734 non-null    float64
 7   Publisher   719 non-null    object 
 8   Skin color  734 non-null    object 
 9   Alignment   734 non-null    object 
 10  Weight      732 non-null    float64
dtypes: float64(2), int64(1), object(8)
memory usage: 63.2+ KB
None


The describe method shows basic statistical characteristics of each numerical feature (int64 and float64 types): number of non-missing values, mean, standard deviation, range, median, 0.25 and 0.75 quartiles. In this case "height" and "weight"

In [270]:
df.describe()

Unnamed: 0      Height      Weight
count  734.000000  734.000000  732.000000
mean   366.500000  102.254087   43.855191
std    212.031837  139.624543  130.823733
min      0.000000  -99.000000  -99.000000
25%    183.250000  -99.000000  -99.000000
50%    366.500000  175.000000   62.000000
75%    549.750000  185.000000   90.000000
max    733.000000  975.000000  900.000000

Return the first rows of the DataFrame `df`.  5 in this example the first the row 5 with df.iloc, and the first 5 rows with df.head

In [271]:
df.iloc[5]


Unnamed: 0                5
name          Absorbing Man
Gender                 Male
Eye color              blue
Race                  Human
Hair color          No Hair
Height                193.0
Publisher     Marvel Comics
Skin color                -
Alignment               bad
Weight                122.0
Name: 5, dtype: object

In [272]:
df.head(5)

Unnamed: 0         name Gender Eye color               Race Hair color  \
0           0       A-Bomb   Male    yellow              Human    No Hair   
1           1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   
2           2     Abin Sur   Male      blue            Ungaran    No Hair   
3           3  Abomination   Male     green  Human / Radiation    No Hair   
4           4      Abraxas   Male      blue      Cosmic Entity      Black   

   Height          Publisher Skin color Alignment  Weight  
0   203.0      Marvel Comics          -      good   441.0  
1   191.0  Dark Horse Comics       blue      good    65.0  
2   185.0          DC Comics        red      good    90.0  
3   203.0      Marvel Comics          -       bad   441.0  
4   -99.0      Marvel Comics          -       bad   -99.0

Select just the 'Gender' and 'Race' columns from the DataFrame `df`.

In [273]:
df.loc[:, ['Gender', 'Race']]

#or 

df[['Gender', 'Race']]

Gender               Race
0      Male              Human
1      Male      Icthyo Sapien
2      Male            Ungaran
3      Male  Human / Radiation
4      Male      Cosmic Entity
..      ...                ...
729  Female              Human
730    Male        Frost Giant
731    Male     Yoda's species
732  Female              Human
733    Male                  -

[734 rows x 2 columns]

Select the data in rows `[3, 4, 8]` *and* in columns `['Gender', 'Race']`.

In [274]:
df.loc[df.index[[130, 40, 18]], ['Gender', 'Race']]

Gender Race
130  Female    -
40   Female    -
18     Male    -

Select only the rows where the number of height is greater than 190.

In [275]:
df[df['Height'] > 190]

Unnamed: 0            name  Gender Eye color               Race  \
0             0          A-Bomb    Male    yellow              Human   
1             1      Abe Sapien    Male      blue      Icthyo Sapien   
3             3     Abomination    Male     green  Human / Radiation   
5             5   Absorbing Man    Male      blue              Human   
10           10      Agent Zero    Male         -                  -   
..          ...             ...     ...       ...                ...   
700         700       Vision II       -       red                  -   
709         709         Warpath    Male     brown             Mutant   
718         718       Wolfsbane  Female     green                  -   
724         724  Wyatt Wingfoot    Male     brown                  -   
730         730            Ymir    Male     white        Frost Giant   

    Hair color  Height          Publisher Skin color Alignment  Weight  
0      No Hair   203.0      Marvel Comics          -      good   441.0  
1      No Hair   191.0  Dark Horse Comics       blue      good    65.0  
3      No Hair   203.0      Marvel Comics          -       bad   441.0  
5      No Hair   193.0      Marvel Comics          -       bad   122.0  
10           -   191.0      Marvel Comics          -      good   104.0  
..         ...     ...                ...        ...       ...     ...  
700    No Hair   191.0      Marvel Comics          -      good   135.0  
709      Black   218.0      Marvel Comics          -      good   158.0  
718     Auburn   366.0      Marvel Comics          -      good   473.0  
724      Black   196.0      Marvel Comics          -      good   117.0  
730    No Hair   304.8      Marvel Comics      white      good   -99.0  

[130 rows x 11 columns]

In [276]:
#Filter just the Columns of name and Height
df[['name', 'Height']][df['Height'] > 190]

name  Height
0            A-Bomb   203.0
1        Abe Sapien   191.0
3       Abomination   203.0
5     Absorbing Man   193.0
10       Agent Zero   191.0
..              ...     ...
700       Vision II   191.0
709         Warpath   218.0
718       Wolfsbane   366.0
724  Wyatt Wingfoot   196.0
730            Ymir   304.8

[130 rows x 2 columns]

In this database where there is no data its "-" so we are using this function to transform this into a null value that python can recognize 


In [277]:
# Strip spaces only from string columns, We use inplace to modify the database
df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)

# Replace '-' with NaN across the entire DataFrame
df.replace('-', np.nan, inplace=True)


In [278]:
df.head(5)

Unnamed: 0         name Gender Eye color               Race Hair color  \
0           0       A-Bomb   Male    yellow              Human    No Hair   
1           1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   
2           2     Abin Sur   Male      blue            Ungaran    No Hair   
3           3  Abomination   Male     green  Human / Radiation    No Hair   
4           4      Abraxas   Male      blue      Cosmic Entity      Black   

   Height          Publisher Skin color Alignment  Weight  
0   203.0      Marvel Comics        NaN      good   441.0  
1   191.0  Dark Horse Comics       blue      good    65.0  
2   185.0          DC Comics        red      good    90.0  
3   203.0      Marvel Comics        NaN       bad   441.0  
4   -99.0      Marvel Comics        NaN       bad   -99.0

In [279]:
#Select the rows where "Skin color" is missing 
df[df['Skin color'].isnull()]

Unnamed: 0             name  Gender Eye color               Race  \
0             0           A-Bomb    Male    yellow              Human   
3             3      Abomination    Male     green  Human / Radiation   
4             4          Abraxas    Male      blue      Cosmic Entity   
5             5    Absorbing Man    Male      blue              Human   
6             6      Adam Monroe    Male      blue                NaN   
..          ...              ...     ...       ...                ...   
727         727      Yellow Claw    Male      blue                NaN   
728         728     Yellowjacket    Male      blue              Human   
729         729  Yellowjacket II  Female      blue              Human   
732         732          Zatanna  Female      blue              Human   
733         733             Zoom    Male       red                NaN   

           Hair color  Height      Publisher Skin color Alignment  Weight  
0             No Hair   203.0  Marvel Comics        NaN      good   441.0  
3             No Hair   203.0  Marvel Comics        NaN       bad   441.0  
4               Black   -99.0  Marvel Comics        NaN       bad   -99.0  
5             No Hair   193.0  Marvel Comics        NaN       bad   122.0  
6               Blond   -99.0   NBC - Heroes        NaN      good   -99.0  
..                ...     ...            ...        ...       ...     ...  
727           No Hair   188.0  Marvel Comics        NaN       bad    95.0  
728             Blond   183.0  Marvel Comics        NaN      good    83.0  
729  Strawberry Blond   165.0  Marvel Comics        NaN      good    52.0  
732             Black   170.0      DC Comics        NaN      good    57.0  
733             Brown   185.0      DC Comics        NaN       bad    81.0  

[662 rows x 11 columns]

 Select the rows where the Gender is a male *and* the height is less than 0.

In [280]:
df[(df['Gender'] == 'Male') & (df['Height'] < 0)]

Unnamed: 0              name Gender Eye color           Race Hair color  \
4             4           Abraxas   Male      blue  Cosmic Entity      Black   
6             6       Adam Monroe   Male      blue            NaN      Blond   
14           14       Alex Mercer   Male       NaN          Human        NaN   
15           15      Alex Woolsly   Male       NaN            NaN        NaN   
18           18  Allan Quatermain   Male       NaN            NaN        NaN   
..          ...               ...    ...       ...            ...        ...   
697         697          Violator   Male       NaN            NaN        NaN   
702         702            Vulcan   Male     black            NaN      Black   
711         711           Watcher   Male       NaN            NaN        NaN   
712         712         Weapon XI   Male       NaN            NaN        NaN   
715         715          Wildfire   Male       NaN            NaN        NaN   

     Height      Publisher Skin color Alignment  Weight  
4     -99.0  Marvel Comics        NaN       bad   -99.0  
6     -99.0   NBC - Heroes        NaN      good   -99.0  
14    -99.0      Wildstorm        NaN       bad   -99.0  
15    -99.0   NBC - Heroes        NaN      good   -99.0  
18    -99.0      Wildstorm        NaN      good   -99.0  
..      ...            ...        ...       ...     ...  
697   -99.0   Image Comics        NaN       bad   -99.0  
702   -99.0  Marvel Comics        NaN      good   -99.0  
711   -99.0  Marvel Comics        NaN      good   -99.0  
712   -99.0  Marvel Comics        NaN       bad   -99.0  
715   -99.0      DC Comics        NaN      good   -99.0  

[147 rows x 11 columns]

Exploring this data we can see some impossible values for height. Since this data is extracted from "scraping" we can evaluate the quality of the data, and that we will be facing some issues. 

Select the rows the Height is between 100 and 200 (inclusive).

In [281]:
df[df['Height'].between(100, 200)]

Unnamed: 0             name  Gender Eye color           Race  \
1             1       Abe Sapien    Male      blue  Icthyo Sapien   
2             2         Abin Sur    Male      blue        Ungaran   
5             5    Absorbing Man    Male      blue          Human   
7             7     Adam Strange    Male      blue          Human   
8             8         Agent 13  Female      blue            NaN   
..          ...              ...     ...       ...            ...   
727         727      Yellow Claw    Male      blue            NaN   
728         728     Yellowjacket    Male      blue          Human   
729         729  Yellowjacket II  Female      blue          Human   
732         732          Zatanna  Female      blue          Human   
733         733             Zoom    Male       red            NaN   

           Hair color  Height          Publisher Skin color Alignment  Weight  
1             No Hair   191.0  Dark Horse Comics       blue      good    65.0  
2             No Hair   185.0          DC Comics        red      good    90.0  
5             No Hair   193.0      Marvel Comics        NaN       bad   122.0  
7               Blond   185.0          DC Comics        NaN      good    88.0  
8               Blond   173.0      Marvel Comics        NaN      good    61.0  
..                ...     ...                ...        ...       ...     ...  
727           No Hair   188.0      Marvel Comics        NaN       bad    95.0  
728             Blond   183.0      Marvel Comics        NaN      good    83.0  
729  Strawberry Blond   165.0      Marvel Comics        NaN      good    52.0  
732             Black   170.0          DC Comics        NaN      good    57.0  
733             Brown   185.0          DC Comics        NaN       bad    81.0  

[449 rows x 11 columns]

In [282]:
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 734 entries, 0 to 733
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Unnamed: 0  734 non-null    int64  
 1   name        734 non-null    object 
 2   Gender      705 non-null    object 
 3   Eye color   562 non-null    object 
 4   Race        430 non-null    object 
 5   Hair color  562 non-null    object 
 6   Height      734 non-null    float64
 7   Publisher   719 non-null    object 
 8   Skin color  72 non-null     object 
 9   Alignment   727 non-null    object 
 10  Weight      732 non-null    float64
dtypes: float64(2), int64(1), object(8)
memory usage: 63.2+ KB
None


Count all names, and count all values within the names

In [283]:
df['name'].count()

734

In [284]:
df['name'].value_counts()

name
Goliath            3
Spider-Man         3
Vindicator         2
Blizzard           2
Speedy             2
                  ..
Fighting Spirit    1
Fin Fang Foom      1
Firebird           1
Firelord           1
Zoom               1
Name: count, Length: 715, dtype: int64

Calculate the mean weight for each Race of superhero in `df`.

In [285]:
df.groupby('Race')['Weight'].mean()

Race
Alien              114.714286
Alpha              -99.000000
Amazon              68.500000
Android            106.444444
Animal             -57.666667
                      ...    
Xenomorph XX121    169.000000
Yautja             234.000000
Yoda's species      17.000000
Zen-Whoberian       77.000000
Zombie             437.000000
Name: Weight, Length: 61, dtype: float64

We can see inconsistent numbers on average, since this DataSet is not yet cleaned, and have atypical numrical values, thats why should calculate in this case the median, Unlike the mean, which is influenced by extreme values (outliers), the median is robust and gives a better representation of the center in skewed distributions.

In [286]:
df.groupby("Race")["Weight"].median()

Race
Alien               98.0
Alpha              -99.0
Amazon              68.5
Android            135.0
Animal             -99.0
                   ...  
Xenomorph XX121    169.0
Yautja             234.0
Yoda's species      17.0
Zen-Whoberian       77.0
Zombie             437.0
Name: Weight, Length: 61, dtype: float64

Sort `df` first by the values in the 'name' in *decending* order, then by the value in the 'Height' column in *ascending* order 

In [287]:
df.sort_values(by=['name', 'Height'], ascending=[False, True])

Unnamed: 0             name  Gender Eye color               Race  \
733         733             Zoom    Male       red                NaN   
732         732          Zatanna  Female      blue              Human   
731         731             Yoda    Male     brown     Yoda's species   
730         730             Ymir    Male     white        Frost Giant   
729         729  Yellowjacket II  Female      blue              Human   
..          ...              ...     ...       ...                ...   
4             4          Abraxas    Male      blue      Cosmic Entity   
3             3      Abomination    Male     green  Human / Radiation   
2             2         Abin Sur    Male      blue            Ungaran   
1             1       Abe Sapien    Male      blue      Icthyo Sapien   
0             0           A-Bomb    Male    yellow              Human   

           Hair color  Height          Publisher Skin color Alignment  Weight  
733             Brown   185.0          DC Comics        NaN       bad    81.0  
732             Black   170.0          DC Comics        NaN      good    57.0  
731             White    66.0       George Lucas      green      good    17.0  
730           No Hair   304.8      Marvel Comics      white      good   -99.0  
729  Strawberry Blond   165.0      Marvel Comics        NaN      good    52.0  
..                ...     ...                ...        ...       ...     ...  
4               Black   -99.0      Marvel Comics        NaN       bad   -99.0  
3             No Hair   203.0      Marvel Comics        NaN       bad   441.0  
2             No Hair   185.0          DC Comics        red      good    90.0  
1             No Hair   191.0  Dark Horse Comics       blue      good    65.0  
0             No Hair   203.0      Marvel Comics        NaN      good   441.0  

[734 rows x 11 columns]

The 'Aligment' column contains the values 'good' and 'bad'. Replace this column with a column of boolean values: 'good' should be `True` and 'bad' should be `False`.

In [288]:
# Safely map only 'good' and 'bad' to boolean values
df['Alignment'] = df['Alignment'].map({'good': True, 'bad': False})

# Handling NaN values in a more future-proof way
df['Alignment'] = df['Alignment'].fillna(np.nan)  # Keeps NaN if you want it

# If you want to fill NaNs with a specific value (e.g., False), explicitly cast the column:
df['Alignment'] = df['Alignment'].fillna(False).astype('bool')

/var/folders/zy/0qz97zt97vj07vpx_gz3j8qh0000gn/T/ipykernel_14538/2286459331.py:8: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df['Alignment'] = df['Alignment'].fillna(False).astype('bool')


In [291]:
df.head(5)

Unnamed: 0         name Gender Eye color               Race Hair color  \
0           0       A-Bomb   Male    yellow              Human    No Hair   
1           1   Abe Sapien   Male      blue      Icthyo Sapien    No Hair   
2           2     Abin Sur   Male      blue            Ungaran    No Hair   
3           3  Abomination   Male     green  Human / Radiation    No Hair   
4           4      Abraxas   Male      blue      Cosmic Entity      Black   

   Height          Publisher Skin color  Alignment  Weight  
0   203.0      Marvel Comics        NaN       True   441.0  
1   191.0  Dark Horse Comics       blue       True    65.0  
2   185.0          DC Comics        red       True    90.0  
3   203.0      Marvel Comics        NaN      False   441.0  
4   -99.0      Marvel Comics        NaN      False   -99.0

Replace Function the 'Gender' column, and change the 'Male' entries to 'M'. and "Female" to "F"

In [318]:
df['Gender'] = df['Gender'].replace({'Male': 'M', 'Female': 'F'})

For each superhero type and each aligment, find the mean age. In other words, each row is an animal, each column is a number of visits and the values are the mean ages (*hint: use a pivot table*).

In [324]:
df.pivot_table(index='Race', columns='Alignment', values='Weight', aggfunc='median')

Alignment           False  True 
Race                            
Alien                97.0   98.0
Alpha                 NaN  -99.0
Amazon                NaN   68.5
Android             146.0   18.0
Animal              -99.0   25.0
Asgardian           230.5  191.0
Atlantean            72.0  115.5
Bizarro             155.0    NaN
Bolovaxian            NaN  324.0
Clone                 NaN  113.0
Cosmic Entity       -99.0    NaN
Cyborg              142.0  173.0
Czarnian            288.0    NaN
Dathomirian Zabrak  -99.0    NaN
Demi-God              NaN   98.5
Demon               104.0  158.0
Eternal             443.0  207.0
Flora Colossus        NaN    4.0
Frost Giant          58.0  -99.0
God / Eternal       -99.0  -99.0
Gorilla             270.0    NaN
Gungan                NaN  -99.0
Human                82.0   66.5
Human / Altered     128.0  103.5
Human / Clone       -99.0    NaN
Human / Cosmic        NaN  -20.0
Human / Radiation   441.0  121.5
Human-Kree            NaN   63.0
Human-Spartoi         NaN   79.0
Human-Vulcan          NaN   81.0
Human-Vuldarian       NaN   95.0
Icthyo Sapien         NaN   65.0
Inhuman               NaN   72.5
Kakarantharaian       NaN   18.0
Korugaran            92.0    NaN
Kryptonian          -99.0   67.5
Luphomoid            83.0    NaN
Maiar               -99.0    NaN
Martian               NaN  135.0
Metahuman           356.0   59.0
Mutant               86.0   58.5
Mutant / Clone        NaN   50.0
New God             135.0    NaN
Neyaphem             67.0    NaN
Parademon           -99.0    NaN
Planet                NaN  -99.0
Rodian              -99.0    NaN
Saiyan               73.0   62.0
Spartoi             -99.0    NaN
Strontian           268.0    NaN
Symbiote            101.5   97.0
Talokite              NaN   54.0
Tamaranean            NaN   71.0
Ungaran               NaN   90.0
Vampire               NaN   -1.0
Xenomorph XX121     169.0    NaN
Yautja              234.0    NaN
Yoda's species        NaN   17.0
Zen-Whoberian         NaN   77.0
Zombie              437.0    NaN

Making a DataFrame easier to work with like here many values are missing **(they are NaN)** 